File: conformance/adapters/python/adapter.py
#!/usr/bin/env python3
"""EHBP conformance adapter (Python). See conformance/adapters/README.md.

Reads one fixture on stdin, runs it through the public ehbp API, prints one
normalized result. All native-error translation lives in map_error.
"""

from __future__ import annotations

import json
import os
import sys

import httpx
from ehbp import (
    Client,
    EHBPTransport,
    ServerIdentity,
    SessionRecoveryToken,
    compute_nonce,
    derive_response_keys,
)
from ehbp.protocol import ENCAPSULATED_KEY_HEADER
from ehbp.errors import (
    CryptoError,
    EHBPError,
    HPKEError,
    InvalidConfigError,
    InvalidInputError,
    KeyConfigMismatchError,
    ProtocolError,
)
from ehbp.protocol import KEYS_PATH, RESPONSE_NONCE_HEADER

HEADER_SUBSET = ("ehbp-response-nonce", "content-length", "transfer-encoding", "content-type")


def main() -> None:
    fx = json.load(sys.stdin)
    res = {"fixture_id": fx["id"], "outcome": "ok", "error_code": None, "status": None,
           "body_hex": None, "passthrough": False, "plaintext_emitted_before_error": False,
           "bytes_emitted_before_error": 0, "native_error": None, "runner": "python"}
    try:
        run(fx, res)
    except EHBPError as err:
        res["outcome"] = "error"
        res["error_code"] = map_error(fx["operation"], err)
        res["body_hex"] = None
        res["native_error"] = f"{type(err).__name__}: {err}"
    except Exception as err:  # non-EHBP failure is still the adapter reporting honestly
        res["outcome"] = "error"
        res["error_code"] = map_error(fx["operation"], err)
        res["native_error"] = f"{type(err).__name__}: {err}"
    json.dump(res, sys.stdout)


def run(fx: dict, res: dict) -> None:
    op = fx["operation"]
    ins = fx.get("inputs", {})
    if op == "derive_keys":
        km = derive_response_keys(b(ins, "exportedSecret"), b(ins, "requestEnc"), b(ins, "responseNonce"))
        res["body_hex"] = (km.key + km.nonce_base).hex()
    elif op == "compute_nonce":
        res["body_hex"] = compute_nonce(b(ins, "nonceBase"), int(ins["seqHex"], 16)).hex()
    elif op in ("decrypt_response", "decrypt_response_streaming"):
        decrypt(fx, res)
    elif op == "token_roundtrip":
        tok = SessionRecoveryToken.from_json(ins["json"])
        res["body_hex"] = (tok.exported_secret + tok.request_enc).hex()
    elif op == "parse_config":
        res["body_hex"] = ServerIdentity.unmarshal_public_config(b(ins, "config")).public_key_bytes().hex()
    elif op == "marshal_config":
        res["body_hex"] = ServerIdentity.from_public_key_bytes(b(ins, "publicKey")).marshal_public_config().hex()
    elif op == "request":
        request(fx, res)
    elif op == "discover":
        Client.discover(discover_target(ins))  # raises on bad content type / status
    elif op in ("reject_reserved_header", "reject_cross_origin", "reject_url_credentials"):
        hardening(op)
    else:
        raise InvalidInputError(f"unknown operation {op}")


def discover_target(ins: dict) -> str:
    return {"bad_ct": os.environ.get("ORACLE_BAD_CT_URL", ""),
            "non200": os.environ.get("ORACLE_NON200_URL", "")}.get(
        ins.get("target"), os.environ["ORACLE_URL"])


def hardening(op: str) -> None:
    base = os.environ["ORACLE_URL"]
    client = Client.discover(base)
    if op == "reject_reserved_header":
        client.request("POST", "/s/echo", body=b"x", headers={ENCAPSULATED_KEY_HEADER: "x"})
    elif op == "reject_cross_origin":
        client.request("GET", "http://other.example/x")
    else:  # reject_url_credentials
        client.request("GET", base.replace("http://", "http://user:pass@") + "/x")


def decrypt(fx: dict, res: dict) -> None:
    ins = fx["inputs"]
    token = SessionRecoveryToken(b(ins, "exportedSecret"), b(ins, "requestEnc"))
    decryptor = token.create_response_decryptor(b(ins, "responseNonce"))
    framed = b(ins, "encryptedResponse")
    segments = split_at(framed, fx.get("chunking")) if fx["operation"].endswith("streaming") else [framed]

    out = bytearray()
    try:
        for seg in segments:
            for chunk in decryptor.push(seg):
                out += chunk
        decryptor.finish()
    except EHBPError:
        res["bytes_emitted_before_error"] = len(out)
        res["plaintext_emitted_before_error"] = len(out) > 0
        raise
    res["body_hex"] = bytes(out).hex()


def request(fx: dict, res: dict) -> None:
    base = os.environ["ORACLE_URL"].rstrip("/")
    keys = httpx.get(base + KEYS_PATH).content
    identity = ServerIdentity.unmarshal_public_config(keys)
    req = fx["request"]
    content = bytes.fromhex(req["body_hex"]) if req.get("body_hex") else None

    client = httpx.Client(transport=EHBPTransport(identity), base_url=base)
    out = bytearray()
    try:
        with client.stream(req["method"], req["path"], content=content,
                           headers=req.get("headers") or {}) as r:
            res["status"] = r.status_code
            res["response_headers"] = subset_headers(r.headers)
            for chunk in r.iter_bytes():
                out += chunk
        res["body_hex"] = bytes(out).hex()
        no_nonce = RESPONSE_NONCE_HEADER.lower() not in {k.lower() for k in (res.get("response_headers") or {})}
        res["passthrough"] = no_nonce and not (200 <= (res["status"] or 0) < 300)
    except EHBPError:
        res["bytes_emitted_before_error"] = len(out)
        res["plaintext_emitted_before_error"] = len(out) > 0
        raise
    finally:
        client.close()


# map_error is the sole native-error -> canonical-code translation.
def map_error(op: str, err: Exception) -> str:
    msg = str(err)
    if isinstance(err, KeyConfigMismatchError):
        return "KEY_CONFIG_MISMATCH"
    if isinstance(err, CryptoError):
        return "AEAD_DECRYPT_FAILED"
    if isinstance(err, HPKEError):
        return "HPKE_SETUP_FAILED"
    if isinstance(err, InvalidConfigError):
        return "UNSUPPORTED_SUITE" if "unsupported" in msg.lower() else "INVALID_KEY_CONFIG"
    if isinstance(err, InvalidInputError):
        return "INVALID_TOKEN" if op == "token_roundtrip" else "INVALID_INPUT"
    if isinstance(err, ProtocolError):
        low = msg.lower()
        if "content type" in low or "returned status" in low:
            return "INVALID_KEY_CONFIG"
        if "missing" in low and "nonce" in low:
            return "MISSING_RESPONSE_NONCE"
        if "multiple" in low:
            return "DUPLICATE_RESPONSE_NONCE"
        if "nonce" in low:
            return "INVALID_RESPONSE_NONCE"
        if "truncated" in low:
            return "FRAMING_TRUNCATED"
        if "exceeds maximum" in low:
            return "CHUNK_TOO_LARGE"
        if "overflow" in low:
            return "SEQUENCE_OVERFLOW"
    return "INVALID_INPUT"


# --- helpers ---

def b(ins: dict, key: str) -> bytes:
    return bytes.fromhex(ins[key])


def split_at(data: bytes, offsets):
    if not offsets:
        return [data]
    segs, prev = [], 0
    for o in offsets:
        if prev < o < len(data):
            segs.append(data[prev:o])
            prev = o
    segs.append(data[prev:])
    return segs


def subset_headers(headers) -> dict:
    out = {}
    for name in HEADER_SUBSET:
        if name in headers:
            out[name] = headers[name]
    return out


if __name__ == "__main__":
    main()

File: conformance/adapters/js/adapter.mjs
// EHBP conformance adapter (JavaScript / Node). See conformance/adapters/README.md.
//
// Reads one fixture on stdin, runs it through the public ehbp API, prints one
// normalized result. All native-error translation lives in mapError.

import {
  Identity,
  createTransport,
  deriveResponseKeys,
  computeNonce,
  decryptResponseWithToken,
  deserializeSessionRecoveryToken,
  hexToBytes,
  bytesToHex,
  KeyConfigMismatchError,
  ProtocolError,
  DecryptionError,
} from '../../../js/dist/esm/index.js';

const HEADER_SUBSET = ['ehbp-response-nonce', 'content-length', 'transfer-encoding', 'content-type'];
const RESPONSE_NONCE_HEADER = 'Ehbp-Response-Nonce';

async function main() {
  const fx = JSON.parse(await readStdin());
  const res = {
    fixture_id: fx.id, outcome: 'ok', error_code: null, status: null, body_hex: null,
    passthrough: false, plaintext_emitted_before_error: false,
    bytes_emitted_before_error: 0, native_error: null, runner: 'js',
  };
  try {
    await run(fx, res);
  } catch (err) {
    res.outcome = 'error';
    res.error_code = mapError(fx.operation, err);
    res.body_hex = null;
    res.native_error = `${err?.name || 'Error'}: ${err?.message || err}`;
  }
  process.stdout.write(JSON.stringify(res));
}

async function run(fx, res) {
  const ins = fx.inputs || {};
  switch (fx.operation) {
    case 'derive_keys': {
      const km = await deriveResponseKeys(hb(ins.exportedSecret), hb(ins.requestEnc), hb(ins.responseNonce));
      res.body_hex = bytesToHex(km.keyBytes) + bytesToHex(km.nonceBase);
      return;
    }
    case 'compute_nonce': {
      const seq = Number.parseInt(ins.seqHex, 16);
      res.body_hex = bytesToHex(computeNonce(hb(ins.nonceBase), seq));
      return;
    }
    case 'decrypt_response':
    case 'decrypt_response_streaming':
      return decrypt(fx, res);
    case 'token_roundtrip': {
      const t = deserializeSessionRecoveryToken(ins.json);
      res.body_hex = bytesToHex(t.exportedSecret) + bytesToHex(t.requestEnc);
      return;
    }
    case 'parse_config': {
      const id = await Identity.unmarshalPublicConfig(hb(ins.config));
      res.body_hex = await id.getPublicKeyHex();
      return;
    }
    case 'marshal_config': {
      const id = await Identity.fromPublicKeyHex(ins.publicKey);
      res.body_hex = bytesToHex(await id.marshalConfig());
      return;
    }
    case 'request':
      return doRequest(fx, res);
    case 'discover':
      await createTransport(discoverTarget(ins));  // throws on bad content type / status
      return;
    case 'reject_reserved_header':
    case 'reject_cross_origin':
    case 'reject_url_credentials':
      res.outcome = 'skipped';
      res.skip_reason = 'js-transport-has-no-request-guards';
      return;
    default:
      throw new Error(`unknown operation ${fx.operation}`);
  }
}

function discoverTarget(ins) {
  if (ins.target === 'bad_ct') return process.env.ORACLE_BAD_CT_URL;
  if (ins.target === 'non200') return process.env.ORACLE_NON200_URL;
  return process.env.ORACLE_URL;
}

async function decrypt(fx, res) {
  const ins = fx.inputs;
  const token = { exportedSecret: hb(ins.exportedSecret), requestEnc: hb(ins.requestEnc) };
  const framed = hb(ins.encryptedResponse);
  const segments = fx.operation.endsWith('streaming') ? splitAt(framed, fx.chunking) : [framed];

  const source = new ReadableStream({
    start(controller) {
      for (const s of segments) controller.enqueue(s);
      controller.close();
    },
  });
  const response = new Response(source, { headers: { [RESPONSE_NONCE_HEADER]: bytesToHex(hb(ins.responseNonce)) } });
  const decrypted = await decryptResponseWithToken(response, token);
  res.body_hex = await drain(decrypted.body, res);
}

async function doRequest(fx, res) {
  const base = process.env.ORACLE_URL.replace(/\/$/, '');
  const transport = await createTransport(base);
  const req = fx.request;
  const init = { method: req.method, headers: req.headers || {} };
  if (req.body_hex) init.body = hb(req.body_hex);
  const response = await transport.request(base + req.path, init);

  res.status = response.status;
  res.response_headers = subsetHeaders(response.headers);
  const noNonce = !(res.response_headers || {})[RESPONSE_NONCE_HEADER.toLowerCase()];
  res.body_hex = await drain(response.body, res);
  res.passthrough = noNonce && !(response.status >= 200 && response.status < 300);
}

// mapError is the sole native-error -> canonical-code translation.
function mapError(op, err) {
  if (err instanceof KeyConfigMismatchError) return 'KEY_CONFIG_MISMATCH';
  if (err instanceof DecryptionError) return 'AEAD_DECRYPT_FAILED';
  const msg = (err?.message || String(err)).toLowerCase();
  if (err instanceof ProtocolError) {
    if (msg.includes('missing') && msg.includes('nonce')) return 'MISSING_RESPONSE_NONCE';
    if (msg.includes('nonce')) return 'INVALID_RESPONSE_NONCE';
    if (msg.includes('truncated')) return 'FRAMING_TRUNCATED';
    if (msg.includes('exceeds maximum')) return 'CHUNK_TOO_LARGE';
    if (msg.includes('no cipher suites')) return 'INVALID_KEY_CONFIG';
    if (msg.includes('cipher suite')) return 'UNSUPPORTED_SUITE';
    return 'INVALID_KEY_CONFIG';
  }
  if (op === 'discover') return 'INVALID_KEY_CONFIG';
  if (op === 'token_roundtrip') return 'INVALID_TOKEN';
  if (op === 'compute_nonce' && msg.includes('sequence')) return 'INVALID_INPUT';
  if (op === 'derive_keys' && msg.includes('must be')) return 'INVALID_INPUT';
  return 'INVALID_INPUT';
}

// --- helpers ---

async function drain(body, res) {
  if (!body) return '';
  const reader = body.getReader();
  const chunks = [];
  let n = 0;
  try {
    for (;;) {
      const { done, value } = await reader.read();
      if (done) break;
      chunks.push(value);
      n += value.length;
    }
  } catch (err) {
    res.bytes_emitted_before_error = n;
    res.plaintext_emitted_before_error = n > 0;
    throw err;
  }
  const out = new Uint8Array(n);
  let off = 0;
  for (const c of chunks) { out.set(c, off); off += c.length; }
  return bytesToHex(out);
}

function splitAt(data, offsets) {
  if (!offsets || offsets.length === 0) return [data];
  const segs = [];
  let prev = 0;
  for (const o of offsets) {
    if (o > prev && o < data.length) { segs.push(data.slice(prev, o)); prev = o; }
  }
  segs.push(data.slice(prev));
  return segs;
}

function subsetHeaders(headers) {
  const out = {};
  for (const name of HEADER_SUBSET) {
    const v = headers.get(name);
    if (v != null) out[name] = v;
  }
  return out;
}

function hb(hex) { return hexToBytes(hex); }

function readStdin() {
  return new Promise((resolve, reject) => {
    let data = '';
    process.stdin.setEncoding('utf8');
    process.stdin.on('data', (c) => (data += c));
    process.stdin.on('end', () => resolve(data));
    process.stdin.on('error', reject);
  });
}

main().catch((err) => { console.error(err); process.exit(2); });
